### ISSUE TITLE
Failure while executing: git clone --depth 1 --branch v0.53.3

### ISSUE THREAD

**Issue by Rhinomcd** [caskroom/homebrew-cask#10041]
I've noticed a few other users had reported this in the past (https://github.com/caskroom/homebrew-cask/issues/2880 seems to be the most relevant), I've read through all the responses and nothing seemed to work for me. I'm fairly confident that there a blatant error on my part that I'm just overlooking. 

I'm trying to install brew-cask, but it fails on 
```
 Failure while executing: git clone --depth 1 --branch v0.53.3 https://github.com/caskroom/homebrew-cask.git /Library/Caches/Homebrew/brew-cask--git
```

Here is what I have tried:
* restarting my computer (https://github.com/caskroom/homebrew-cask/issues/5966)
* reinstalling homebrew
* reinstalling ruby(https://github.com/caskroom/homebrew-cask/issues/250#issuecomment-21777530)
* reinstalling git
* brew edit brew-cask( https://github.com/caskroom/homebrew-cask/issues/1003#issuecomment-39394090)
* manually executing ```git clone --depth 1 --branch v0.53.3 https://github.com/caskroom/homebrew-cask.git /Library/Caches/Homebrew/brew-cask--git```
    * this checks out successfully 

Verbose install
```
TWY018:~ $ brew install -vd brew-cask
/usr/local/Library/brew.rb (Formulary::FormulaLoader): loading /usr/local/Library/Taps/caskroom/homebrew-cask/brew-cask.rb
==> Installing brew-cask from caskroom/homebrew-cask
/usr/local/Library/Homebrew/build.rb (Formulary::FromPathLoader): loading /usr/local/Library/Taps/caskroom/homebrew-cask/brew-cask.rb
/usr/local/Library/Homebrew/build.rb (Formulary::FormulaLoader): loading /usr/local/Library/Formula/git.rb
==> Cloning https://github.com/caskroom/homebrew-cask.git
git --git-dir /Library/Caches/Homebrew/brew-cask--git/.git status -s
Removing invalid repository from cache
git clone --depth 1 --branch v0.53.3 https://github.com/caskroom/homebrew-cask.git /Library/Caches/Homebrew/brew-cask--git
Error: Failed to download resource "brew-cask"
Failure while executing: git clone --depth 1 --branch v0.53.3 https://github.com/caskroom/homebrew-cask.git /Library/Caches/Homebrew/brew-cask--git
/usr/local/Library/Homebrew/utils.rb:151:in `safe_system'
/usr/local/Library/Homebrew/download_strategy.rb:629:in `clone_repo'
/usr/local/Library/Homebrew/download_strategy.rb:119:in `fetch'
/usr/local/Library/Homebrew/resource.rb:109:in `fetch'
/usr/local/Library/Homebrew/resource.rb:80:in `stage'
/usr/local/Library/Homebrew/formula.rb:766:in `stage'
/usr/local/Library/Homebrew/formula.rb:406:in `brew'
/usr/local/Library/Homebrew/build.rb:114:in `install'
/usr/local/Library/Homebrew/build.rb:185:in `<main>'
HTWY018:~ $ 
```

My Brew config:
```
HTWY018:~ $ brew config
HOMEBREW_VERSION: 0.9.5
ORIGIN: https://github.com/Homebrew/homebrew
HEAD: ccd981d823fc506bbf99811023a2ef1c792d3eb7
Last commit: 3 hours ago
HOMEBREW_PREFIX: /usr/local
HOMEBREW_CELLAR: /usr/local/Cellar
CPU: quad-core 64-bit sandybridge
OS X: 10.10.2-x86_64
Xcode: 6.2
CLT: 6.2.0.0.1.1424975374
Clang: 6.0 build 600
X11: N/A
System Ruby: 2.0.0-p481
Perl: /usr/bin/perl
Python: /usr/bin/python
Ruby: /usr/local/bin/ruby => /usr/local/homebrew/Cellar/ruby/2.2.1/bin/ruby
Java: 1.8.0_31
HTWY018:~ $ 

````

which git
```
HTWY018:~ $ which git
/usr/local/bin/git
```
which ruby
```
HTWY018:~ $ which ruby
/usr/local/bin/ruby
```

If there is any more information required please let me know.  I appreciate any support I can get.  I've been bashing my head into this off and on for around 2 weeks now.

**Comment by Amorymeltzer** [comment]
Hi @Rhinomcd, sorry you've had so much trouble!  This has clearly come up a lot, and going through the issues here and elsewhere (homebrew itself) I'm with you - it's unclear what should work.

Question: If you run 
````
cd `brew --prefix`
`````
and then do a `git status`, what's it look like?  Does `brew update` fail too?  If so, try:
````
cd `brew --prefix`
git fetch origin
git reset --hard origin/master
brew update
````
I think this error results from a `brew` command cutting out midway, whether loss of internet or, in my case, "Oh damn didn't mean to do that ^C ^C ^C!"

**Comment by Rhinomcd** [comment]
git status
```
HTWY018:local $ git status
On branch master
nothing to commit, working directory clean
```

brew update
```
HTWY018:local $ brew update
Updated Homebrew from d223841c to d223841c.
==> Updated Formulae
homebrew/versions/openssl098		homebrew/versions/openssl101
```

I appreciate the help! Still the same Failure while executing error though. :/

**Comment by Amorymeltzer** [comment]
Okay, same thing but for your `homebrew-cask` repo?  Usually in `$(brew --repository)/Library/Taps/caskroom/homebrew-cask/`

**Comment by Rhinomcd** [comment]
```
HTWY018:~ $ cd  $(brew --repository)/Library/Taps/caskroom/homebrew-cask/
HTWY018:homebrew-cask $ git status
On branch master
Your branch is up-to-date with 'origin/master'.
nothing to commit, working directory clean
```
ran the following anyway
```
git fetch origin
git reset --hard origin/master
brew update
```
Still the same Failure while executing error.

**Comment by Stocki** [comment]
I have exactly the same error ... any news on that issue?

**Comment by Rhinomcd** [comment]
@Stocki, Unfortunately, I haven't been able to progress past what has been said in this thread so far.  Hopefully we can isolate what causes this and find a fix.

**Comment by JimiSpire** [comment]
Got the same failure here.

**Comment by Rhinomcd** [comment]
I'm sad to say this issue seems to have resolved itself when I tried again this morning... I wish I could Isolate what the fix was, but it seems like magic to me.  Here is my current brew-config so people can see if they can find what was fixed.

```
HTWY018:~ $ brew config
HOMEBREW_VERSION: 0.9.5
ORIGIN: https://github.com/Homebrew/homebrew
HEAD: cfceeb9a42042b8e2d4a17f6a09895800f87a634
Last commit: 7 hours ago
HOMEBREW_PREFIX: /usr/local
HOMEBREW_CELLAR: /usr/local/Cellar
CPU: quad-core 64-bit sandybridge
OS X: 10.10.2-x86_64
Xcode: 6.2
CLT: 6.2.0.0.1.1424975374
Clang: 6.0 build 600
X11: N/A
System Ruby: 2.0.0-p481
Perl: /usr/bin/perl
Python: /usr/bin/python
Ruby: /usr/local/bin/ruby => /usr/local/Cellar/ruby/2.2.2/bin/ruby
Java: 1.8.0_45
```

**Comment by swyckoff** [comment]
I had a very similar issue. 

TL;DR: I'm ultimately responsible for removing my /usr/bin/ruby symlink at some point. However it would be good to have a better error message than "Failure while executing: git clone --branch v0.53.3 https://github.com/caskroom/homebrew-cask.git /Library/Caches/Homebrew/brew-cask--git" to figure this out.

Hopefully seeing my struggle will help someone else. Here is a summary of how I figured it out / fixed it (after trying everything above in this issue): 

➜ projects brew config 
HOMEBREW_VERSION: 0.9.5
ORIGIN: https://github.com/Homebrew/homebrew.git
HEAD: f1a2ed10b2b175fa8b518a5a5c2d6687e72e03b5
Last commit: 2 hours ago
HOMEBREW_PREFIX: /usr/local
HOMEBREW_CELLAR: /usr/local/Cellar
CPU: 8-core 64-bit haswell
OS X: 10.10.3-x86_64
Xcode: 6.3
CLT: 6.3.0.0.1.1428348375
Clang: 6.1 build 602
X11: 2.7.7 => /opt/X11
System Ruby: 2.0.0-p481
Perl: /usr/bin/perl
Python: /usr/bin/python
Ruby: /Users/sympl/.rvm/rubies/ruby-2.2.2/bin/ruby
Java: 1.8.0_45

➜ homebrew-cask git:(master) ✗ brew install -vd brew-cask
/usr/local/Library/brew.rb (Formulary::FormulaLoader): loading /usr/local/Library/Taps/caskroom/homebrew-cask/brew-cask.rb
==> Installing brew-cask from caskroom/homebrew-cask
/usr/local/Library/Homebrew/build.rb (Formulary::FromPathLoader): loading /usr/local/Library/Taps/caskroom/homebrew-cask/brew-cask.rb
/usr/local/Library/Homebrew/build.rb (Formulary::FormulaLoader): loading /usr/local/Library/Formula/git.rb
/usr/local/Library/Homebrew/build.rb (Formulary::FormulaLoader): loading /usr/local/Library/Formula/ruby.rb
==> Cloning https://github.com/caskroom/homebrew-cask.git
git clone --branch v0.53.3 https://github.com/caskroom/homebrew-cask.git /Library/Caches/Homebrew/brew-cask--git
Error: Failed to download resource "brew-cask"
Failure while executing: git clone --branch v0.53.3 https://github.com/caskroom/homebrew-cask.git /Library/Caches/Homebrew/brew-cask--git
/usr/local/Library/Homebrew/utils.rb:152:in `safe_system'
/usr/local/Library/Homebrew/download_strategy.rb:631:in `clone_repo'
/usr/local/Library/Homebrew/download_strategy.rb:121:in `fetch'
/usr/local/Library/Homebrew/resource.rb:109:in `fetch'
/usr/local/Library/Homebrew/resource.rb:80:in `stage'
/usr/local/Library/Homebrew/formula.rb:876:in `stage'
/usr/local/Library/Homebrew/formula.rb:507:in `brew'
/usr/local/Library/Homebrew/build.rb:107:in `install'
/usr/local/Library/Homebrew/build.rb:179:in `<main>'

#### let’s see what version of git it uses... ####
➜ vi /usr/local/Library/Homebrew/download_strategy.rb
ADD: puts `which git` 

def clone_repo
     puts `which git` # add this guy
     safe_system 'git', *clone_args
     cached_location.cd do
          safe_system "git", "config", "homebrew.cacheversion", cache_version
          update_submodules if submodules?
     end
end

#### again: `brew install brew-cask` ####
➜ projects brew install -vd brew-cask
/usr/local/Library/brew.rb (Formulary::FormulaLoader): loading /usr/local/Library/Taps/caskroom/homebrew-cask/brew-cask.rb
==> Installing brew-cask from caskroom/homebrew-cask
/usr/local/Library/Homebrew/build.rb (Formulary::FromPathLoader): loading /usr/local/Library/Taps/caskroom/homebrew-cask/brew-cask.rb
/usr/local/Library/Homebrew/build.rb (Formulary::FormulaLoader): loading /usr/local/Library/Formula/git.rb
/usr/local/Library/Homebrew/build.rb (Formulary::FormulaLoader): loading /usr/local/Library/Formula/ruby.rb
==> Cloning https://github.com/caskroom/homebrew-cask.git
/usr/local/Library/ENV/4.3/git 
git clone --branch v0.53.3 https://github.com/caskroom/homebrew-cask.git /Library/Caches/Homebrew/brew-cask--git
Error: Failed to download resource "brew-cask"
Failure while executing: git clone --branch v0.53.3 https://github.com/caskroom/homebrew-cask.git /Library/Caches/Homebrew/brew-cask--git

#### See "/usr/local/Library/ENV/4.3/git "? ####

➜ projects cd /usr/local/Library/ENV/4.3
➜ 4.3 git:(master) ✗ ls -al git
lrwxr-xr-x 1 sympl admin 10 Apr 20 11:06 git -> ../scm/git
➜ 4.3 git:(master) ✗ ls -al ../scm/git
-rwxr-xr-x 1 sympl admin 1167 Apr 20 11:06 ../scm/git
➜ 4.3 git:(master) ✗ ../scm/git --version
zsh: ../scm/git: bad interpreter: /usr/bin/ruby: no such file or directory 

➜ 4.3 git:(master) ✗ sudo ln -s /System/Library/Frameworks/Ruby.framework/Versions/Current/usr/bin/ruby /usr/bin/ruby

➜ 4.3 git:(master) ✗ ../scm/git --version
git version 2.3.5 #looks good.

➜ projects brew install -vd brew-cask
/usr/local/Library/brew.rb (Formulary::FormulaLoader): loading /usr/local/Library/Taps/caskroom/homebrew-cask/brew-cask.rb
==> Installing brew-cask from caskroom/homebrew-cask
/usr/local/Library/Homebrew/build.rb (Formulary::FromPathLoader): loading /usr/local/Library/Taps/caskroom/homebrew-cask/brew-cask.rb
/usr/local/Library/Homebrew/build.rb (Formulary::FormulaLoader): loading /usr/local/Library/Formula/git.rb
/usr/local/Library/Homebrew/build.rb (Formulary::FormulaLoader): loading /usr/local/Library/Formula/ruby.rb
==> Cloning https://github.com/caskroom/homebrew-cask.git
/usr/local/Library/ENV/4.3/git
git --version
git version 2.3.5
git clone --branch v0.53.3 https://github.com/caskroom/homebrew-cask.git /Library/Caches/Homebrew/brew-cask--git
Cloning into '/Library/Caches/Homebrew/brew-cask--git'...
remote: Counting objects: 135049, done.
remote: Compressing objects: 100% (12/12), done.
remote: Total 135049 (delta 4), reused 0 (delta 0), pack-reused 135037
Receiving objects: 100% (135049/135049), 40.71 MiB | 6.92 MiB/s, done.
Resolving deltas: 100% (89413/89413), done.
Checking connectivity... done.
Note: checking out 'e65c40bb2bb02d7aa26f3d0ed6217862f2436055'.

You are in 'detached HEAD' state. You can look around, make experimental
changes and commit them, and you can discard any commits you make in this
state without impacting any branches by performing another checkout.

If you want to create a new branch to retain commits you create, you may
do so (now or later) by using -b with the checkout command again. Example:

git checkout -b new_branch_name

git config homebrew.cacheversion 0
==> Checking out tag v0.53.3
==> Cleaning
==> Finishing up
ln -s ../Cellar/brew-cask/0.53.3/bin/brew-cask brew-cask
ln -s ../../../Cellar/brew-cask/0.53.3/share/man/man1/brew-cask.1 brew-cask.1
/usr/local/Library/Homebrew/postinstall.rb (Formulary::FromPathLoader): loading /usr/local/Library/Taps/caskroom/homebrew-cask/brew-cask.rb
==> Summary
🍺 /usr/local/Cellar/brew-cask/0.53.3: 2421 files, 9.5M, built in 14 seconds

4.3 git:(master) ✗ brew cask
==> We need to make Caskroom for the first time at /etc/Caskroom
==> We'll set permissions properly so we won't need sudo in the future

Great Success!

**Comment by goldfish0506** [comment]
@swyckoff 
This is my issue too, sincerely appreciate !

**Comment by toralux** [comment]
I had the very same issue, thanks a lot for helping my fixing this problem! :)
Make sure to put backtick-characters between "puts" and "which git", and to put it in correct "def clone_repo" block (you will find it several times in "download_strategy.rb"

**Comment by swyckoff** [comment]
@toralux thanks, I fixed my markdown, should be clearer now. @goldfish0506 np!

**Comment by rmurcek** [comment]
Thanks @swyckoff ! Had the same issue brew installing rbenv from git
<pre><code> brew install rbenv --HEAD </code></pre>
and this saved me.

**Comment by dhawal55** [comment]
Awesome!! This fixed it. Thank you.

**Comment by ishabo** [comment]
Running 
```
ln -s /System/Library/Frameworks/Ruby.framework/Versions/Current/usr/bin/ruby /usr/bin/ruby 
```
before 
```
brew install -vd brew-cask
```
Solved the issue for me!

**Comment by javfindly** [comment]
I was using RVM. I've created a symlink into /usr/bin/ruby and it started to work. Thank you very much for the descriptive error explanation
